<h2>Welcome to NanoEngineer-1 Version 1.1.1</h2>

NanoEngineer-1 (NE1) is an open-source (GPL) 3D multi-scale modeling and simulation program for nano-composites with special support for structural DNA nanotechnology. It features an easy-to-use interactive 3D graphical user interface for designing and modeling large, atomically precise composite nanosystems.

NanoEngineer-1 Version 1.1.1 is developed by <A HREF="http://www.nanorex.com">Nanorex, Inc.</A> in the hope that it will accelerate the development of advanced nanosystems.

<h2>System Requirements</h2>

<h3>Minimum</h3>

<li> 2.5 GHz Intel Pentium 4 CPU or equivalent
<li> Windows XP, Mac OS X 10.4.x, Linux
<li> 1 Gb RAM
<li> 100 Mb available disk space
<li> OpenGL compatible 3D video card

<h3>Recommended</h3>

<li> 3 GHz Intel Core 2 CPU or equivalent
<li> Windows XP, Mac OS X 10.4.x, Linux
<li> 2 Gb RAM
<li> 100 Mb available disk space
<li> OpenGL compatible 3D video card with FBO and VBO support (<A HREF="http://www.delphi3d.net/hardware/extsupport.php?extension=GL_EXT_framebuffer_object"> this web page</A> gives a good list of recommended video cards.)

<h2>Tested Platforms</h2>

NanoEngineer-1 Version 1.1.1 has been tested on:

<li> Windows XP
<li> Mac OS X 10.4.9 (Tiger), and 10.5.2 (Leopard)
<li> Kubuntu 7.10 (Gutsy)

Limited testing has been done on:

<li> Windows Vista - it seems to work fine on Vista Business, and with some successes on Vista Home

<h2>Installation Instructions</h2>
<TITLE>NanoEngineer-1 Version 1.1.1 Readme</TITLE>

<h3>Windows</h3>

Run the NanoEngineer-1 self-extracting installer "NanoEngineer-1_1.1.1.exe".

<h3>Mac OS X</h3>

Run the NanoEngineer-1 self-extracting installer "NanoEngineer-1_1.1.1.dmg".

Note: If during installation, the message: "You cannot continue. There is nothing to install." appears, navigate to <your hard drive>/Library/Receipts/ and delete any files that start with "NanoEngineer-1". Then empty the Trash and run the installer again.

<h3>Linux and Developers</h3>

NanoEngineer-1 Version 1.1.1 must be compiled from sources for Linux users and developers. Download the NanoEngineer-1 Version 1.1.1 source distribution "NanoEngineer-1_1.1.1.tar.gz" and follow the instructions found in the following steps:

<ol>
	<li> <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=NE1_Build_Requirements"> NE1 Build Requirements</A> - Before building NE1, check that the requirements for your platform are met. This page lists all the libraries necessary to build NE1, including some notes on library installation.
	<li> <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=NE1_Build_and_Install"> NE1 Build and Install</A>
</ol>
<h2>Getting Started with NE1</h2>

The best way to get started using NanoEngineer-1 is:

<ol>
	<li> Peruse this <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Introduction_to_the_NanoEngineer-1_User_Interface"> Introduction to the NanoEngineer-1 User Interface</A>.
	<li> Then work through some of the tutorials on our <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Tutorials"> Tutorials wiki page</A>.
</ol>

<h2>Extensions for NE1</h2>

NE1 makes use of a number of free extensions and companion applications, including:

<li> GROMACS an extremely high performance, open source, and cross-platform package used to perform molecular dynamics simulation of systems with hundreds to millions of particles.
<li> Open Babel, used to import/export different chemistry file formats to/from NE1.
<li> QuteMolX, a high-quality molecular visualization system that can be "launched and loaded" from NE1. Very cool!
<li> POV-Ray, which creates high-quality renderings of the current model with a single click of the mouse.
<li> MegaPOV (for Mac OS X), which creates high-quality renderings of the current model with a single click of the mouse. POV-Ray is also required.
<li> GNUPlot, used to plot the results of molecular dynamics simulations using NanoDynamics-1, NE1's own molecular dynamics engine for nanomechanical simulations. It is installed with NE1 for Windows and Mac.

For more details on all extensions and links to download pages, visit: <A HREF="http://nanoengineer-1.com/dev1/index.php?option=com_content&task=view&id=31&Itemid=86"> NanoEngineer-1 Extensions</A>.

Note that the <b>NanoEngineer-1 Application Suite</b> includes all of the above extensions and companion applications in one convenient installer. You can find it at <A HREF="http://www.nanorex.com/"> www.nanorex.com</A>.

<h2>What's New in NanoEngineer-1 Version 1.1.1</h2>

<h3>New commands</h3>
<ol>
<li> New <b>Lighting Scheme</b> command and property manager for editing lights and saving Lighting Scheme ''favorites''.
<li> New <b>Help > NanoEngineer-1 Tutorials...</b> menu item.
</ol>

<h3>Enhancements</h3>
<ol>
<li> DNA related enhancements:
<ul>
<li> A DNA duplex can be drawn (inserted) on a specified <b>Plane</b>.
<li> Improved DnaSegment ''resize handles'' to ease interactive resize operations.
<li> The <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Cursor_text">cursor text</A> is larger and easier to read.
<li> The <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Cursor_text">cursor text</A> follows the cursor vertically while resizing segments.
<li> The <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Cursor_text">cursor text</A> remains within the Model View area, even when the cursor is along the right edge.
<li> DnaStrand and DnaSegment colors can be changed from the property manager.
<li> New <b>View > Display > DNA Cylinder</b> menu item.
<li> New <b>Edit > DNA Display Style</b> menu item.
<li> Default color of 3' and 5' arrowheads are green.
<li> <b>Extrude</b> command places new copies of strands and segments into their original DnaGroup(s).
</ul>
<li> Plane related enhancements:
<ul>
<li> Grid lines can be displayed/changed on plane surface.
<li> Images are saved/restored between NE1 sessions.
<li> Select a <b>Plane</b> by clicking anywhere inside it.
</ul>
<li> New <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Preferences_dialog">Preferences dialog</A> options:
<ul>
<li> <b>General</b> page: <b>Offset factor for pasting objects</b> group box.
<li> <b>Model View</b> page: <b>Display confirmation corner</b> check box.
</ul>
</ol>

<h3>Fixed Bugs</h3>
<ol>
<li> DNA related bug fixes:
<ul>
<li> DnaSegment property manager displays the correct <b>Bases per turn</b> value.
<li> DnaSegment property manager displays the correct number of base pairs.
<li> The <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Cursor_text">cursor text</A> displays the correct length delta when resizing a DnaSegment.
<li> Resizing a DnaSegment no longer adds one base pair when finished.
<li> The ''Bases per turn'' and ''Duplex rise'' values in the <b>Insert DNA</b> property manager now honor values in the  <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Preferences_dialog">Preferences dialog</A>.
</ul>
<li> Fixed property manager font size inconsistencies.
<li> Open Babel export now executes correctly.
<li> Open Babel export no longer zooms the screen after export.
<li> Miscellaneous screen and menu layout problems concerning the switch to Qt 4.3.5.
<li> Modified the file selection box extension choices for Open Babel export/import so that file extensions are displayed on Mac.
</ol>

<h3> Miscellaneous </h3>
<ol>
<li> The default <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Model_View">Model View</A> background color is ''Evening Sky''.
<li> Improved color contrast between the <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Model_View">Model View</A> background color and line color of region selection polyline (rectangle/lasso), DNA lines and Nanotube lines.
<li> <b>Insert > Grid Plane</b> has been deprecated. Use <b>Insert > Plane</b> to add a plane with a grid.
<li> Support for IOS file import/export.
<li> NanoEngineer-1 is now built with Qt 4.3.5.
</ol>
----

See the <b>New Features Archives</b> below for information on previous versions.

<h2>Known Issues and Bugs</h2>

Currently, NanoEngineer-1 and its installers should work properly on MacOSX 10.4+. Due to a switch to a newer version of the installer system, the NanoEngineer-1 folder may be installed directly into the toplevel folder of the root partition, rather than into its intended location of /Applications/Nanorex. During testing, this error did not occur on versions of MacOSX 10.4 and above. If this occurs, just move the entire NanoEngineer-1 folder to its intended location using the Finder. The application itself may not work on MacOSX 10.3 as it did before, due to a crash apparently related to our use of a newer version of Qt (4.3.5). You may access a current copy of NE1 Known Bugs on-line at:
<A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=NE1_Known_Bugs">NE1 Known Bugs</A>

<h2>Copyright and Licensing</h2>

NanoEngineer-1 Copyright 2004-2008 Nanorex, Inc.

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

<h2>New Features Archives</h2>

<h3>What's New in NanoEngineer-1 Version 1.1.0 (6/4/2008):</h3>

<ol><li> User Interface Modifications
<ul>
<li> Resize multiple DnaSegments at once.
<li> DNA cursor text display options in a new PM group box. Display options (as checkboxes): Bases, Turns, Length, Angle.
<li> DNA and Nanotube cursor text should auto-switch b/w Angstroms (0-10 A) and nanometers ( > 10 A).
<li> atom overlap indicator - quick access to a toggle for it in the UI
<li> Edit DNA Display Style property manager that allows the user to edit and preview DNA display options.
<li> Order DNA property manager modifications.
<li> Make Crossovers command (initial implementation, for PAM3 model)
<li> Colored wire-frames as an option for hover highlighting and selection.
<li> Cleaned up the Run Dynamics and Minimize Energy dialogs.
</ul>
<li> Other new features
<ul>
<li> NE1 installers associate files with the .mmp extension with NE1 for all platforms.
<li> GROMACS 3.3.3 build (double precision)
<li> Better support for languages which have characters which do not show up in the ASCII character set
<li> Added the ability to display an image on a reference plane
<li> Added support for stereo image displays.
<li> Added support for fog, accessible from Color Scheme PM and the Preferences dialog.
</ul>
<li> Fixed Bugs
<ul>
<li> Hover highlighting not working after mouse wheel zooming.
<li> reduce minimum size of User Prefs Dialog widget
<li> fixed bug in breaking a PAM5 crossover bond
<li> other minimizer issues listed at PAM model tasks for v1.1.0
<li> fixed ND1 PAM3 minimize for crossovers
<li> fixed bug in strand list widget selection or update, exposed or caused by ghost bases being in same DnaGroup as their axis
<li> spontaneous PAM5 conversion on duplex resize
<li> misc. necessary features to speed development and testing of PAM force field
<li> fix major bug in saving a PAM5 ring duplex
<li> revise Minimize Energy in the following ways:
<li>* always "Minimize All" by default; nonempty selection enables (but does not check) checkbox for "Minimize Selection"
<li>* Minimize Selection passes *all* atoms (not just those near enough to the selection), anchoring all unselected atoms
<li> leave out PAM bondpoints instead of writing H in place of them
<li> PAM sugar atoms aren't connected when a dna segment is resized by 1 basepair
<li> Cancel button to abort reading a huge file doesn't work
</ol>

<h4>PAM5 and PAM3 DNA Force Field Improvements (for v1.1.0)</h4>

Both the PAM5 and PAM3 (pseudo-atom reduced model, 5 and 3 pseudo-atom structure) DNA force fields have undergone refinement. <br /><br />

<b>1. PAM5-Specific Improvements</b><br /><br /> This version of NE1 contains an alpha release of the PAM5 (pseudo-atom model, 5 pseudo-atom structure) force field for the modeling of DNA-based nanostructures.This force field is being provided prior to publication for testing and benchmarking purposes.While further refinement is likely to follow, the current version has been found to be useful in the modeling of several classes of DNA nanostructures and is not expected to undergo significant modification. <br /><br /> The force constants generated from the parameterization process are the result of comparisons with mutually inconsistent sources.The bending stiffness values suffer most from the available selection of data sets used for the force constant development.Consequently, the relative energies of strained structures should not (yet) be considered meaningful.The beta version will fit stretch, bend, and twist modulii (as well as the stretch-twist cross term and several others) based on a combination of experimental and molecular dynamics (published and currently “in-house”) results. <br /><br /> More information about the PAM5 force field, including a description of the interhelical potential and example structures, is available at:<br /><br />[[PAM3_and_PAM5_Model_Descriptions]] <br /><br />

<b>2. Interhelical Description</b><br /><br /> The screened Coulombic (Yukawa) potential used to model the interhelical interactions in PAM5 has been implemented as an additional energy minimization option for the PAM3 force field.The Yukawa potential is used in place of the Coulombic potential on the backbone pseudo-atoms.In PAM3, the charged pseudo-atoms mark the sugar positions.In PAM5, the charged pseudo-atoms mark the phosphate positions. <br /><br /> The constants in the Yukawa potential have the following default settings (which can be adjusted to new default settings in the sim-params.txt file): <br />

<li> Counter-ion Charge: 2.0 (Mg<sup>2+</sup> counter-ion)
<li> Counter-ion Molarity: 0.02 M
<li> System Temperature: 298.0 K
<li> Dielectric Constant: 78.5 (water)

<br /> The Yukawa potential has the following general influences on structure optimizations: <br />

<li> Increased separation (“bowing”) between neighboring duplexes (duplex-duplex repulsion)
<li> Increased separation between free duplex ends (those not connected by crossovers) at tight convergence (EndRMS < 5.0 pN)
<li> Curvature at the corners of three-dimensional DNA structures (a result of local repulsive interactions at junction points)

The screened Coulombic (Yukawa) potential is always repulsive as currently implemented.In atomistic simulations containing discrete counter-ions, duplexes can be bridged by these counter-ions to provide local binding stability.This leads to duplex loops with smaller average separations (“bowing”) and duplex ends in crossover junctions that do not separate.As a screened Columbic potential applied to the phosphate (PAM5) or backbone sugar (PAM3) pseudo-atoms, the Yukawa potential provides no such local binding between duplexes. <br /><br />

<b>3. PAM3/PAM5 Energy Minimization Issues and Procedures</b><br /><br /> A general procedure for modeling DNA nanostructures using the PAM5 force field is as follows: <br /><br />

<p><li> Draw the components using the PAM3 structures in the “Build DNA” Mode
<p> Perform an Energy Minimization using PAM3
<li> With or Without Electrostatics (the Yukawa potential)
<li> With or Without Neighbor Searching (regeneration of the atom pair table for the calculation of electrostatic interactions.This slows down the energy minimization significantly but leads to more accurate structures)
<p> Confirm the Optimized Structure minimized correctly
<li> If not, adjust the structure and perform an additional Energy Minimization
<p> Convert the PAM3 structure to PAM5 with the “PAM3 to PAM5” Button
<p> Perform an Energy Minimization using PAM5
<li> With or Without Electrostatics (the Yukawa potential)
<li> With or Without Neighbor Searching (regeneration of the atom pair table for the calculation of electrostatic interactions.This slows down the energy minimization significantly but leads to more accurate structures)
<p> Confirm the Optimized Structure minimized correctly
<li> If not, adjust the structure and perform an additional Energy Minimization

<br> The PAM3 force field contains fewer pseudo-atoms than PAM5 and relatively stiff bending terms.This makes energy minimization faster with PAM3, especially when the starting structure is significantly different than the optimized (equilibrium) structure.As a general rule, the minimization procedure for a new structure should begin with PAM3 prior to the more accurate, but more time consuming, PAM5 minimization. 

<h3>What's New in NanoEngineer-1 1.0.1 (5/6/2008):</h3>

<p> User Interface Modifications
<li> New selection/highlighting/drag behavior in Build Dna and its sub-commands..
<li> Edit DnaStrands and DnaSegments on single-click.
<li> Option to render 3' and 5' arrowheads/spheres in different color.
<li> Convert bonds in "Build Chunks" command.
<li> New Dna seletion/highlighting and drag behavior in all modes. See Dna Selection and Movement perations.
<p> Fixed Bugs
<li> DNA Cylinder display style: Axis should not be displayed on ssDNA segments.
<li> Double-clicking strand does not select mate strands in MT. Hiding all strands only shows double-clicked strand as hidden in MT.
<li> undo/redo doesn't restore selectedness of DnaStrands
<li> Joining strands doesn't always pull in all necessary chunks (except for the dna updater).
<li> Closing or "tearing-off" history makes the energy minimize fail.
<li> Tooltip shows the Windows hotkey even on Mac.
<li> Fix Default path for all NE1 plug-ins.
<li> Zoom to proper scale (i.e. 50A half height) when editing a duplex the first time.
<li> Button images are missing from the sequence editor on Mac, an installer issue.
<li> Strand resize handle color is not updated for new strand.
<li> Atoms not rendered when global display style is "Dna Cylinder".
<li> Default settings for DNA Cylinder display style.
<li> Length of sequence unknown when typing into sequence editor.
<li> Tooltips and Help docs should include keyboard shortcuts.
<li> Bug when a strand is lengthened by 1 strand base atom
<li> Hitting enter in sequence editor inserts newline (which appears to clear the sequence) and inserts an "N" into the mates field.
<li> Mao 3-point star has discrepancies.
<li> SDN structure clean up in part library.

<h3>What's New in NanoEngineer-1 1.0.0 (4/25/2008):</h3>

<p> Core User Interface Operations Support
<li> A new DNA data model has been implemented that makes use of a unique DNA structure checking mechanism.
<p> User Interface Improvements and Additions
<li> DNA Strand Operations
<li>* New handles for resizing segments and strands along the axis.
<li>* Nick/break/join strands - now you can break a strand or join two strands via a break/join command on the toolbar.
<li>* Create a crossover in one simple operation by selecting nucleobases from adjacent strands and invoking the context menu.
<li> DNA Segment Operations
<li>* Change the length of segment
<li>** Constrained to central axis
<li>** By adding base pairs
<li>* Rotate a segment about its central axis
<li>* Translate a segment along its central axis
<li> DNA Duplex Operations
<li>* Create a single DNA duplex by selecting two points on a plane
<li>** The duplex is shown while you're selecting the second point and can be displayed in either ladder or ribbon display styles.
<li>** The duplex adheres to intelligent horizontal and vertical "snap-to-grid" rules allowing for rapid, horizontal and vertical placement.
<li>** Text next to the mouse cursor dynamically updates the number of base-pairs and the length of the duplex as you change the rubberband line during DNA duplex creation.
<li> Sequence Editor
<li>* Load sequences from a text file, paste them in, or even type them.
<li>* Unique find/replace feature lets you quickly find and replace sub-sequences.
<li>* Easily switch between 5'-3' and 3'-5' orientation.
<li>* Save sequences to a separate text file.
<li> Peptide Generator
<li>* This experimental module can create atomistic models of polypeptide chains of arbitrary sequence and secondary structure.
<li>* The polypeptide chain can be built step-by-step by choosing consecutive amino acids from a set of the standard 20 naturally occurring amino acids.
<li>* Local geometry of the built chain can be chosen from a set of predefined secondary structure geometries (alpha helix, 3-10-helix, pi-helix, or beta-strand) or explicitly defined by specifying a pair of phi/psi angles.
<li>* A sequence of the peptide is visualized in a sequence viewer window and it is colored by current secondary structure type.
<li>* Note:
<li>** Generated structures can't be minimized or simulated in a reliable way because the ND1 force field currently lacks certain atom types (e.g. amide-type nitrogen). The generated geometry is idealized, but after minimization it becomes wrong.
<li>** The structure is not visualized as it is being built, you need to press the "Preview" button to see it.
<li>** The structure can't be modified after it has been generated, nor can the sequence be edited.
<li> Miscellaneous Improvements and Additions
<li>* DNA generator now supports PAM3 and PAM5 models, but <b>not</b> atomistic models.
<li>* A new rainbow-effect cylinder display style is now available for viewing reduced model PAM DNA structures.
<li>* A new nanotube generator - create a single-walled carbon nanotube by selecting two points on a plane.
<li>* Toolbar and dockwidget positions can now be saved between sessions.
<li>* Option to maintain the aspect ratio while resizing a plane. The ratio is updated automatically (i.e. anytime the plane's width and/or height are changed).
<li>* The bond offset sphere checkbox in Extrude property manager is back.
<li>* Mirror feature enhancements - it can now mirror everything but the planes.
<li>* Triple-click any PAM atom to select all connected PAM atoms.
<li>* New paste mode and related enhancements
<li>** New paste mode which can be accessed using menu Edit > Paste from clipboard... or corresponding button in the standard toolbar. You can return to the previous mode by hitting hitting the 'Done' or 'Escape' key.
<li>** Clipboard groups are now 'pastable' (so those are also included in the clipboard groupbox of Paste Mode)
<li>** Edit > Paste or Ctrl + V does a 'single-shot' paste operation. It pastes the most recently copied item. The item placement position is slightly modified compared to the original copied item.
<li>** Improvements to group pasting via paste mode. Now the groups are pasted at a closer place to where you double-clicked.
<li>* New part library mode - the part library can be accessed in any mode by using Insert > Part from partlib... menu. You can return to the previous mode by hitting the 'Done' or 'escape' key.
<li>* When a single atom is selected, a new groupbox in the Build Atoms Property Manager, Selection Options, becomes active and shows the x, y, z coords of the selected atom. You can then use the spinboxes to move the atom precisely.
<li>* NE1 can now be shown in full-screen mode (only menus and rulers appear around the work area) and semi-full-screen mode (includes all the toolbars as well.)
<p> Structural DNA Nanotech (SDN) Part Library Additions
<li> Several DNA double-crossovers such as DAE, DAO, DPE, DPON, DPOW, DX Tile, DX+J Tile, JX2, and PX65
<li> Several SDN structures such as Mao's 3- and 6-point stars, Rothemund's star origami, Pierce's walker, Turberfield's tetrahedron, and Shih's octahedron.
<p>+ PAM5 and PAM3 Reduced DNA Models
<li> The original PAM5 has undergone further refinements to improve the realism of minimized SDN structures
<li> A new PAM3 model has been added for simplified creation, and clear, attractive display of SDN motifs
<p>+ QuteMol<sub>X</sub> Integration
<li> QuteMol<sub>X</sub> is our branch of QuteMol. The branch allows us to continue to improve the code, and better integrate it with NE1.
<li> QuteMol<sub>X</sub> now matches the display mode, background color, and atom coloring of NE1 when either launched from NE1, or when opening PDB files exported from NE1.
<p>+ GROMACS Integration
<li> GROMACS has been added as an energy minimization engine. Structures can be minimized in the foreground, or in a separate background GROMACS process. The location of all the GROMACS input/output files, and the identifier (PID) for the GROMACS process is printed in the History text area at the bottom of NE1. An application called NanoVision-1, which is currently in development and not yet available, will be able to attach to, monitor, and cleanly abort the spawned GROMACS processes.
<li> HDF5 support was added to GMX for writing and viewing trajectory files, but won't be fully realized until NanoVision-1 is available.
<p>+ Notable Bug Fixes
<li> The NE1 process now reliably ends upon exit of NE1.
<li> POV-Ray scene fixed - bonds are now the correct color when a chunk is displayed in ball-and-stick style, and colored.
<li> PAM reduced DNA models are now rendered properly in QuteMolX.
<li> Fixed the "Rotate as a unit" checkbox in the Rotate groupbox.
<li> Moiety copies are now placed closer together when extruding to a ring as final product.
<li> Inserting a part with a hotspot from the parts library now works.
<li> Newly read mmp files are no longer immediately marked as changed.
<li> Added a mouse wheel governor for Mac users - using the mouse wheel to zoom is now manageable.

<h3>What's New in NanoEngineer-1 0.9.1 (7/01/2007):</h3>

<p> UI Improvements
<li> New <A HREF="http://www.nanoengineer-1.net/mediawiki/index.php?title=Confirmation_Corner"> Confirmation Corner</A>.
<li> Command Manager includes custom flyout toolbars for all commands
<li> Property Manager includes helpful messages for <b>Plane</b>, <b>Build Atoms</b>, <b>Extrude</b> and <b>Translate/Rotate</b> commands
<li> <b>Build Atoms</b> includes the new <b>Cut Bonds</b> bond tool for deleting bonds between two atoms
<li> Splitter between the left channel (i.e. the Model Tree/Property Manager) and the 3D graphics window can be resized
<p> DNA Generator
<li> Specify DNA duplex by length (in Angstroms)
<li> Fixed some bugs involving the Strand Sequence text edit widget (in the Property Manager)
<p> <A HREF="http://qutemol.sourceforge.net/"> QuteMol</A>
<li> NE1 writes and QuteMol (0.4.1) reads and renders bondpoints
<li> NE1 now writes the PDB REMARK records needed for QuteMol (a future version) to render:
<li>* Initial view orientation
<li>* Background color
<li>* Launch display style (launch = "synchronize with NE1 at startup")
<li>* Launch tubes/licorice cylinder "thickness"
<li>* Launch ball and stick cylinder "thickness"
<li>* Chunks' display style and color  Be warned, however, that QuteMol demands high-end graphics hardware. Also:
<li>* QuteMol runs well on the "Pro" series of Mac products, (i.e. Macbook Pro and Mac Pro), but it can hang other Macs (i.e. MacMini, MacBook, and some other PowerPC based Macs). QuteMol has not been known to hang Windows-based computers.
<li>* Check <A HREF="http://www.delphi3d.net/hardware/extsupport.php?extension=GL_EXT_framebuffer_object"> this list</A> of supported graphics cards and drivers.
<li>* If your computer includes a graphics card on this list but still isn't rendering properly, then be sure to update your graphics card driver.
<p> Reference Plane (<b>Insert > Reference Geometry > Plane</b>)
<li> New option <b>Offset to a Plane</b> creates a new plane from an offset to an existing plane
<li> Edges are aligned with the screen when <b>Parallel to Screen</b> option is selected
<li> Works with <b>Mirror</b> command
<li> Works for all <b>Translate</b> commands (i.e. <b>To XYZ Position</b>)
<li> Preview improvements
<li> Many resizing bugs fixed
<p> Other notable bug fixes:
<li> The current working directory is set to the last directory in which a file was opened or saved. This is persistent between seesions
<li> The graphitic Nitrogen atomic hybrid can be deposited and transmutes correctly
<li> <b>Simulation > Play Movie</b> is enabled only when there is a valid movie
<li> Fixed some button text overlapping and layout problems in <b>Build Atoms</b> Property Manager

<h3>What's New in NanoEngineer-1 v0.9 (6/1/2007):</h3>

<p> Completely ported GUI from Qt 3.x to Qt 4.x
<li> NanoEngineer-1 Alpha 9 is now GPL'd for all platforms, including Windows
<p> Improved Graphical User Interface, improved ease of use
<li> New Property Manager presents users with options for current commands
<li> New Command Manager Toolbar provides easy access to most common commands
<li> Improved selection and highlighting heuristics
<li> Better, more intuitive interactive graphics for translating and rotating selected components
<li> Interactive zooming on cursor focus point
<p> Special support for Structural DNA Nanotechnology
<li> DNA "reduced model" allows easy design of complex DNA structures
<li> Collection of DNA motifs included in part library:
<li>* Basic motifs (i.e. holliday junctions, double crossovers, triple crossovers, etc.)
<li>* Star structures
<li>* DNA tiles
<li>* DNA origami

<h3>What's New in NanoEngineer-1 v0.8 (7/13/2006):</h3>

<p> New Structure Generators
<li> DNA Generator - creates custom DNA structures
<li>* B-DNA and Z-DNA
<li>* Single or double stranded
<li> Enhanced Nanotube Generator
<li>* Carbon or Boron Nitride nanotubes
<li>* Multi-walled nanotubes
<li>* Twisted option
<li>* Bend option
<li>* Z and XY distortions options
<li> Heterojunction
<li>* Generates and connects two nanotubes of arbitrary size.
<li>* It uses using a C++ port of CoNTub, an algorithm for connecting two arbitrary carbon nanotubes.
<li> Graphene Generator
<li>* Creates a sheet of graphene given width and height dimensions.
<p> POV-Ray Scene and Ray Trace scene
<li> These provide a convenient way to preview the results of a POV-Ray scene rendering in NanoEngineer-1. To use this feature, you should install POV-Ray or MegaPOV and enable it as a plug-in.
<p> Support for adding comments to the model
<p> New Chunk display
<li> Chunks can be rendered in a new Cylinder display mode. (Experimental feature)
<p> Simulator dialog
<li> Options to specify how often to update the screen during the simulation.
<p> Adjust All and Adjust Selection
<li> Quickly adjusts positions of atoms and bonds to make geometry of structures more realistic.
<p> Minimize Energy
<li> New dialog for providing fine grain control of structure minimizations.
<p> Simulator:
<li> Torsion and out-of-plane terms in the simulator. (Although these terms are included, please note that this still has some known bugs and need more work. For example, minimize returning high values for some structures or taking long time to finish)
<p> Infer bonds, when reading PDB files that don't have any bonds
<p> Build Mode improvements:
<li> Bond types can be changed using context menu or bond tools, even when this requires changing atom types or removing bondpoints.
<li> The bond order of a bondpoint can be changed using bond tools.
<li> Automatic repositioning of bondpoints when dragging connected atoms has been improved

<h3>What's New in NanoEngineer-1 v0.7 (4/24/2006):</h3>

<p> Undo/Redo:
<li> NanoEngineer-1 now includes rich Undo/Redo functionality
<li> Automatic checkpointing can be enabled/disabled by selecting 'Edit > Automatic Checkpointing'
<li> 'Edit > Clear Undo Stack' clears all Undo checkpoints to free system memory
<p> 'Select Atoms Mode' has been merged with 'Build Mode'
<p> Build Mode:
<li> Significantly improved interactive mouse controls for building, selecting, moving and editing structures and other objects. To learn more about the new mouse controls, see the 'Build Mode' section of the Mouse Control documentation by selecting 'Help > Mouse Controls...'
<li> Hover highlighting now supports bonds and jigs (not just atoms)
<li> Atoms, bonds and jigs have context menus for common commands
<li> Atom Selection Filter, which filters the selection of one or more atom types when performing atom selection operations
<li> Transmute atoms
<li> New Library tab in the MMKit allows 3D preview and insertion of parts from the NanoEngineer-1 library
<li> 'Water' selection filter can be enabled/disabled from the dashboard (disabled by default)
<li> Build mode is the new default startup mode
<p> On-Line Wiki Help:
<li> Context sensitive help using the F1 key
<li> Users can contribute to and/or edit the On-Line Wiki Help documentation
<p> Molecular Dynamic Simulation:
<li> Improved force field. Formal bend parameters have replaced guessed values for all single bond combinations
<li> Warnings emitted when parameters are used that are not determined by quantum calculations
<li> Warnings when bonds exceed reasonable limits during simulation
<li> Watch minimize and MD simulations in real time
<li> Minimize and MD simulations can be stopped at any time, leaving atoms at their current positions
<li> Improved minimization algorithm
<p> View Menu/Toolbar:
<li> 'Set View Normal To' sets the view to a new view based on the normal vector of the current selection
<li> 'Set View Parallel To' sets the view to a new view based on the parallel vector of the current selection
<li> 'Rotate View 180' rotates the current view 180 degrees around the vertical axis
<li> 'Rotate View +90' rotates the current view 90 degrees around the vertical axis
<li> 'Rotate View -90' rotates the current view -90 degrees around the vertical axis
<li> 'Save Named View' saves the current view as a custom view in the Model Tree. Using the context menu of the Named View icon in the Model Tree, the view can be restored
<p> Single Walled Carbon Nanotubes:
<li> SWCNTs can be added to the model by selecting 'Insert > Nanotube' and specifying the chirality and length
<p> Graphics:
<li> In general, rendering speed and quality are improved
<li> 'View Animation', which animates the model between the current view and the destination view. This helps maintain model/view orientation when switching to a standard view (i.e. Front, Top, Home, etc.)
<li> 'Help > Graphic Card Info...' displays details about the system's graphics card
<p> New Jigs:
<li> 'Measure Distance Jig' displays the distance between the nuclei and the VdW radii of two atoms
<li> 'Measure Angle' displays the angle made by three atoms
<li> 'Measure Dihedral' displays the dihedral angle of a four atom sequence
<li> 'ESP Image' allows the user to visualize the electrostatic potential of points on the face of a square 2D surface. Nano-Hive's MPQC ESP Plane plug-in is used to calculate the electrostatic potential
<li> 'Gridplane' allows the user to insert a 2D square or SiC grid into the part to help measure distances or fuse chunks of SiC
<li> 'Atom Set' allows the user to group and easily select arbitrary sets of atoms in the part
<p> Nano-Hive:
<li> ESP Image plug-in (tested on Windows only)
<p> Fuse Chunks Mode:
<li> Allows user to fuse two or more chunks together using one of two methods:
<li># Make Bonds - bondpoints between chunks will form bonds when they are near each other
<li># Fuse Atoms - atoms between chunks will be fused when they overlap each other
<p> Movie Player:
<li> New 'Loop' option on dashboard restarts the movie when it reaches the end
<li> 'POV-Ray Series' supports the 'Skip' value in the dashboard so that the skipped frames are not saved as POV-Ray files
<p> New Preferences (Edit > Preferences):
<li> Display Compass Labels
<li> View Animation
<li>* Animation speed can be changed using a slider control in the 'Edit > Preferences | General' dialog
<li>* View Animation can be enabled/disabled
<li> Minimization: Watch in real-time
<li> Bondpoint Highlighting Color
<li> Ball and Stick Atom Scale
<li> Ball and Stick Bond Scale
<li> CPK Atom Scale
<li> Level of Detail
<li> Bond Line Thickness
<li> Startup Mode
<li> Default Mode
<li> Display Mode (for each Mode)
<li> Build Mode Defaults settings, including Autobond, Water, Highlighting, and Select Atoms of Deposited Object
<li> Additional Lighting parameters, including Color, Specular value and light position
<li> Material Specular Properties
<li> Nano-Hive Plug-in
<li> Undo settings, including:
<li>* Restore View when Undoing Structural Changes
<li>* Automatic Checkpoints (default setting)
<li> Window Position and Size Save button
<p> Miscellaneous:
<li> 'Select > Expand Selection' selects any atom that is a neighbor of a currently selected atom
<li> 'Select > Contract Selection' deselects any atom that is a neighbor of a non-picked atom or has a bondpoint
<li> 'Modify > Merge' operation extended for atom selection (i.e. Build Mode)
<li> 'File > Recent Files' facilitates opening the most recently used files.

----

<h3>What's New in NanoEngineer-1 v0.0.6 (8/17/2005):</h3>

<p> Atomic hybridization and higher order bond types:
<li> Support for the following atomic hybrid orbitals:
<li>* Carbon: sp3, sp2, sp
<li>* Nitrogen: sp3, sp2, sp, graphitic
<li>* Oxygen: sp3, sp2
<li>* Sulfur: sp3, sp2
<li>* Hydrogen: s
<li> Support for the following bond types:
<li>* single
<li>* double
<li>* triple
<li>* aromatic
<li>* graphitic
<li>* carbomeric*  * carbomeric bonds are an order-2.5 bond, sometimes used in carbomers. Alpha 6 supports drawing and valence checking of carbomeric bonds, but they are simulated and stored in the mmp file as aromatic bonds. This means carbomeric bonds can't be stored and reloaded from a mmp file, which is a known bug we'll fix for Alpha 7.
<li> Build Mode:
<li>* New Molecular Modeling Kit (MMKit) used for selecting elements and atomic hybridizations
<li>* Bonds are selectable and can be deleted using CTRL+ Left-click (or CMD + Left-click on MacOS X)
<li>* Bonds can be changed using new sub-modes on the dashboard
<li>* Clipboard integration with the MMKit for copying and pasting molecules
<li>* Keyboard accelerators for element and atomic hybridization selections
<li> Select Atoms Mode:
<li>* Support for copying/pasting selected atoms
<li>* Improved dashboard for selecting (and transmuting) atoms
<li>* Selection Filter now works for Select Connected and Select Doubly
<li>* Select Connected and Select Doubly have been added to the dashboard
<li>* Select Doubly much faster
<li> New GAMESS client interface:
<li>* Energy calculations
<li>* Optimizations
<li>* Supports PC GAMESS on Windows
<li>* Supports GAMESS-US on Linux/MacOSX
<li> Jigs:
<li>* Selection of Jigs in the 3D workspace (graphic area)
<li>* Copying and pasting of jigs (together with their chunk/atoms)
<li>* New model tree context menu item to select all atoms of any Jig or set of Jigs
<li>* New model tree context menu item to disable/enable Jigs, which determines whether the Jig is used during a simulation
<li>* New model tree context menu item for Rotary/Linear Motors "Align to Chunk", which relaigns the motor to the first chunk it is attached to
<li>* New model tree context menu item for Rotary/Linear Motors "Recenter on Atoms", which recenters the motor to the atoms it is attached to
<li> Molecular Dynamics Simulator:
<li>* Revised Simulator Setup dialog
<li>* Support for Linear motor
<li>* Jigs can be disabled (and enabled) in the model tree via their context menu
<li>* MD supported for atomic hybridizations and higher order bonds (see list above) with the following exceptions:
<li>** Still no torsion; this will make systems with double or aromatic/graphitic bonds act less stiff than they really should. The parameters for aromatic and graphitic bonds are interpolated rather than measured directly
<li>** The simulator does not handle hydrogen bonds or dative bonds
<li>** The parameters for stretches with fluorine and bends with boron or a double bond to nitrogen are estimated rather than computed
<li>** The whole set may need an overall correction factor of some kind. The main variable, absolute energy numbers from Gamess, is self canceling since we're only using relative energies, and we used the same level of theory for all the runs. Still there may be some systematic correction that is needed
<li>** A more sophisticated fitting procedure, and support for electrostatics, is in the works for Alpha 7
<li> New Preferences dialog (Edit > Preferences):
<li>* atom highlight color
<li>* free valence color
<li>* free valence hotspot color
<li>* bond highlight color
<li>* bond vanes/ribbons color
<li>* bond stretch color
<li>* CPK cylinder color
<li>* higher order bond display modes: multiple cylinders, vanes, or ribbons
<li>* show valence erros on/off
<li>* turn on/off bond letters for double, triple, aromatic and graphitic bonds
<li>* background color (solid or gradient)
<li>* compass location
<li>* compass on/off
<li>* origin axis on/off
<li>* point of view axis on/off
<li> Move Chunks:
<li>* New dashboard dropbox includes the following options:
<li>** Translate
<li>** Rotate X (by specified angle "Theta")
<li>** Rotate Y (by specified angle "Theta")
<li>** Rotate Z (by specified angle "Theta")
<li> Movie Player:
<li>* Significant speed improvement
<li>* Circular arrow on end of rotary motor shaft rotates with motor
<li>* New "POV-Ray Series" save option that allows the user to dump the current movie as a series of .pov files (which can then be povrayed by the user and collected into an mpeg offline)
<li> Model Tree:
<li>* New model tree context menu item "Delete all clipboard items"
<li>* New model tree context menu item "Create new empty clipboard item"
<li> Opposite View:
<li>* New command (View > Opposite View) that flips the current view to the opposite side
<li> Align to Common Axis:
<li>* Now aligns selected chunks to the first selected chunk's axis (instead of the average axis of all selected chunks)

----

<h3>What's New in NanoEngineer-1 v0.0.5 (5/2/2005):</h3>

<p><li> Minimize Selection: Will minimize the current selection only.
<p>Cookie Cutter:
<li> Support for Lonsdaleite (hexagonal diamond)
<li> Modify selection curve(s) for any cookie layer
<li> Diamond surface orientation displayed in the status bar
<li> Rotate lattice grid by increments, around the POV (point of view) axis.
<li> Abort/restart the current selection using the Esc key
<li> Cookie layers are displayed in different colors
<li> Option to display selection as spheres, similar to CPK mode, but with bonds drawn as white lines.
<li> Option to change the grid line color
<li> Option to display the rest of the model
<li> New shape selection options (triangle, rectangle, hexagon, square and diamond)
<p>Clipboard:
<li> Selected items are now displayed and editable in Graphics Area.
<li> Drag and drop works reliably between the Model Tree and the Clipboard
<p>Move Chunks:
<li> Translate selected chunks by a specified X, Y, Z offset
<li> Move selected chunk(s) to an absolute position
<li> Dashboard options offer constrained rotation and translation along the X, Y, or Z axis.
<li> Modifier keys activate constrained rotation and translation along the X (X key), Y (Y key) or Z (Z key) axis.
<li> Modifier keys limit interactive movement to rotation (R key) or translation (T key)
<p><li> Fuse Chunks:
<li> New tool for interactively connecting open bonds between two or more chunks
<li> Option to merge connected chunks into a single chunk
<li> Fine control (using modifier keys) of chunk movement while finding and highlighting bondable pairs of atom between chunks.
<p><li> Modify > Invert: Inverts the atom positions of the selected chunk(s) around their center in space
<li> Select Atoms: New Atom Selection Filter available on dashboard
<p>Simulator:
<li> Improvements in minimizer, returning the best position from run
<li> Multiple runs may improve position further
<li> Open bonds are treated as if they were hydrogenated.
<li> More informative history messages for the user after a simulation or minimization completes.
<p><li> New Select Menu: Includes all options from the Select Toolbar
<li> Display > Show Invisible Atoms: Resets the display of any invisible atoms in the selected chunk(s).
<li> Display > Reset Chunk Display: Resets the display of all atoms in the selected chunk(s).
<li> Plot Tool: Changed GNUplot filename suffix from "-trace.plt" to "-plot.txt"
<p><li> Removed Element Selector from Modify toolbar.
<p><li> Added Element Selector to the Build Atoms dashboard.
<p><li> Lighting Tool: Added Restore Defaults button.
<p><li> Changed Modify > Weld to Modify > Merge
<p><li> Hide/Unhide state of Model Tree nodes are saved in the MMP file and restored when reopening the part.
<p><li> Open/close state of Model Tree groups are saved in MMP file and restored when reopening the part.
<p>Jigs:
<li> Jigs can be disabled using the Model Tree context menu, or by dragging them to the Clipboard; disabled jigs remain attached to the same atoms, even when saved to and reloaded from an MMP file, but have no effect on the simulator until they are re-enabled.
<li> Jigs can be located anywhere in the Model Tree, and this location is not altered when saving the MMP file.

----

<h3>What's New in v0.0.4 (3/14/2005):</h3>

<p> Element Selector:
<li> removed controls for changing element colors.
<li> Refined buttons to be easier to read, especially on MacOSX.
<p> Element Color Settings:
<li> new control panel for changing element colors.
<li> accessed by "Display > Element Color Settings..."
<li> element colors saved to nE-1 preferences database
<li> element colors can be saved to a text file in human readable format
<li> load element colors from a file
<li> can restore default nE-1 element colors
<p> Plot Tool:
<li> new control panel for plotting simulator trace file output using GNUplot.
<li> buttons on control panel open trace and plot files in a text editor.
<li> plot files can be edited by hand and replotted using "Plot" button
<p>
<li> Simulator trace files now created next to DPB trajectory file, with the name dpb_filename-trace.txt
<li> Simulator: new toolbar/menu contains Simulator, Movie Player and Plot Tool.
<li> Lighting: New Lighting control panel for changing lighting settings.

----

<h3>What's New in v0.0.3 (2/28/2005):</h3>

<li> Select > Doubly: The cursor displays the "Wait Cursor" (hourglass) while Select > Doubly is working.
<li> Model Tree: Improved Drag and Drop behavour in the Model Tree widget.
<li> Clipboard: Fixed some bugs related to the Clipboard. More stable now.
<li> Movie Player: Program no longer displays errors when selecting Done while movie is playing.
<li> Element Selector: Users can change element colors for all supported element types.
<li> Element Selector: Users can load/save custom element colors from/to a file.
<li> Extrude Tool: The bond-offset spheres are turned off by default. They can be turned on from the dashboard.
<li> Extrude Tool: The bond-offset spheres are rendered in a manner to clue the user on which configurations will produce the most bonds.

----

<h3>What's New in v0.0.2 (2/14/2005):</h3>

<li> Simulator works on MacOSX.
<li> Fixed bug in cad that limited the simulator to no more than 940 frames on Windows.

----

<h3>What's New in v0.0.1 (2/2/2005):</h3>

<p> Everything

